### PR TITLE
Pr92x L1T Calo (EG, MET, Tau)  2017 emulation of startup (up to l1t integration v95.15)

### PR DESCRIPTION
**Pull request by rekovic**
92x L1T integration ( replaces #19104)

This contains l1t-integration developments upto tag v95.16. 

Total of 57 files (31 C++, 26 expert level configuration .py files).

Included features:

- EG: new extended H/E, plus 2nd isolation WP (aka 2016)
- MET
   - Emulator bug fix when getting address using ieta for MET PUS LUT, make consistent counting in emulator/fw (0/1).
   - saturate when there is a saturated ECAL or HCAL TP + Saturate MET Demux (fix).
- Tau adding trimming + shape veto dummy LUT
- EG/Tau  clustering filtering fix in emulator to describe fw: explore TT30 the veto procedure when finding maxima.
- caloParams (default 2017_v1_8)
  - caloParams2017_v1_7 to reflect current online conditions.
  - **caloParams **2017_v1_8** (same as v1_7 except layer1HFScaleFactors are now lower, as no longer divided by 0.7).
   - caloParams_2017_v1_8_2, enabling the 2nd iso EG WP.

- new L1TReEmul workflows: 
  - L1TReEmulMCFromRAWSimEcalTP
  - L1TReEmulMCFromRAWSimHcalTP
  - L1TReEmulMCFromRAWSimCalTP. 
  - L1TReEmulMCFrom90xRAWSimHcalTP, suitable for 90x MC samples which contain simHcalUnsuppressedDigis.

### CODE REVIEW

**Comment by rekovic** [comment]
please test

**Comment by rekovic** [comment]
please test

**Comment by rekovic** [comment]
+1

File: EventFilter/L1TRawToDigi/plugins/implementations_stage2/CaloTowerUnpacker.cc [hunk 1]
```diff
@@ -18,7 +18,7 @@ namespace stage2 {
      unsigned int bxid = block.amc().getBX();
 
      // handle offset between BC0 marker and actual BC0...
-     if( (tmt-1) != ((bxid-1+3)%9) ) return true;
+     //if( (tmt-1) != ((bxid-1+3)%9) ) return true;
```
Comment (davidlange6) [inline EventFilter/L1TRawToDigi/plugins/implementations_stage2/CaloTowerUnpacker.cc]: hi @rekovic - perhaps this comment needs updating? The functionality it describes I guess is now removed?

File: EventFilter/L1TRawToDigi/plugins/implementations_stage2/L1TStage2Layer2Constants.cc [hunk 2]
```diff
@@ -1,6 +1,8 @@
 
 #include "L1TStage2Layer2Constants.h"
 
+signed int l1t::stage2::layer2::fedId = 1360;
```
Comment (davidlange6) [inline EventFilter/L1TRawToDigi/plugins/implementations_stage2/L1TStage2Layer2Constants.cc]: why should this be signed?

File: EventFilter/L1TRawToDigi/plugins/implementations_stage2/MPUnpacker.cc [hunk 3]
```diff
@@ -21,7 +21,7 @@ namespace stage2 {
      unsigned int amc  = block.amc().getAMCNumber();
      unsigned int bxid = block.amc().getBX();
      //     if( (amc-1) != (bxid-1)%9 ) return true;
-     if( (amc-1) != ((bxid-1+3)%9) ) return true;   // temporary measure!
+     //if( (amc-1) != ((bxid-1+3)%9) ) return true;   // temporary measure!
```
Comment (davidlange6) [inline EventFilter/L1TRawToDigi/plugins/implementations_stage2/MPUnpacker.cc]: please elaborate on the "temporary measure"

File: L1Trigger/Configuration/python/customiseReEmul.py [hunk 4]
```diff
@@ -168,6 +168,42 @@ def L1TReEmulMCFromRAW(process):
             process.simOmtfDigis.srcCSC             = cms.InputTag('simCscTriggerPrimitiveDigis','MPCSORTED')
     return process
 
+def L1TReEmulMCFromRAWSimEcalTP(process):
+    L1TReEmulMCFromRAW(process)
+    if stage2L1Trigger.isChosen():
```
Comment (davidlange6) [inline L1Trigger/Configuration/python/customiseReEmul.py]: please avoid isChosen(). maybe hasattr works?

File: L1Trigger/Configuration/python/customiseSettings.py [hunk 5]
```diff
@@ -1,6 +1,50 @@
 import os.path
 import FWCore.ParameterSet.Config as cms
 
+def L1TSettingsToCaloStage2Params_2017_v1_9_inconsistent(process):
```
Comment (davidlange6) [inline L1Trigger/Configuration/python/customiseSettings.py]: seems time to implement a helper function? eg
def L1TSettingsToCaloStage2Params(process,str):
   process.load("L1Trigger.L1TCalorimeter.caloStage2Params_"+str+"_cfi")
  return process

then the user calls
process=L1TSettingsToCaloStage2Params(process,"2017_v1_9_inconsistent")

File: L1Trigger/L1TCalorimeter/interface/CaloParamsHelper.h [hunk 6]
```diff
@@ -214,6 +236,7 @@ namespace l1t {
 
     l1t::LUT* tauIsolationLUT() { return &pnode_[tauIsolation].LUT_; }
     l1t::LUT* tauIsolationLUT2() { return &pnode_[tauIsolation2].LUT_; }
+    l1t::LUT* tauTrimmingShapeVetoLUT() {return &pnode_[tauTrimmingShapeVeto].LUT_; }
```
Comment (davidlange6) [inline L1Trigger/L1TCalorimeter/interface/CaloParamsHelper.h]: could (should) this be const?

File: L1Trigger/L1TCalorimeter/interface/CaloStage2Nav.h [hunk 7]
```diff
@@ -39,13 +40,90 @@ namespace l1t{
...
+	      if(iEta_tmp+offset<=-etaMax) return -etaMax;
+	      return iEta_tmp+offset;
+	    }
+	  if(offset>=0)
+	    {
+	      if(iEta_tmp+offset>=etaMax) return etaMax;
+	      return iEta_tmp+offset;
+	    }
+	}
+      else if(iEta>0 && iEta<etaBoundaryHF)
```
Comment (davidlange6) [inline L1Trigger/L1TCalorimeter/interface/CaloStage2Nav.h]: to get here, iEta<etaBoundraryHF - no need to recheck

File: L1Trigger/L1TCalorimeter/plugins/L1TStage2InputPatternWriter.cc [hunk 8]
```diff
@@ -265,53 +271,82 @@ void
...
+  LogDebug("L1TDebug") << "Read " << nEvents_ << " events" << std::endl;
+  LogDebug("L1TDebug") << "Writing " << nOutFiles << " files" << std::endl;
+  LogDebug("L1TDebug") << "Output directory: ./" << outDir_ << "/" << std::endl;
+  
+  //files
+  std::vector< std::ofstream > outFiles(nOutFiles);
+    
+  //make output files and write to them
+  for(uint itFile=0; itFile<nOutFiles; ++itFile){
+    std::stringstream outFilename;
```
Comment (davidlange6) [inline L1Trigger/L1TCalorimeter/plugins/L1TStage2InputPatternWriter.cc]: no real need to use a stringstream here.

File: L1Trigger/L1TCalorimeter/python/caloStage2Params_2017_v1_6_2_cfi.py [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+
+#Vector with 6 parameters for eta bin, from low eta to high
+# 1,0,1,0,1,1 gives no correction
+# must be in this form as may require > 255 arguments
+
+# Or vector with 8 parameters, which caps correction value below given pT
+# as 6 parameters, but last two are max correction and L1 pT below which cap is applied, respectively
+
+jetCalibParamsVector = cms.vdouble()
+jetCalibParamsVector.extend([
```
Comment (davidlange6) [inline L1Trigger/L1TCalorimeter/python/caloStage2Params_2017_v1_6_2_cfi.py]: please add a comment about the source of these (and other) magic numbers

File: EventFilter/L1TRawToDigi/plugins/implementations_stage2/CaloTowerUnpacker.cc (hunk 1 above)
Comment (rekovic) [inline EventFilter/L1TRawToDigi/plugins/implementations_stage2/CaloTowerUnpacker.cc]: Yes, the functionality is moved to the proper place, form Calo- and MP-unpacker to  L1TRawToDigi. The unpacker is smarter and makes the TMT node check, and ensures that only the MP that processed a given BX is unpacked.
Will remove obsolete commented lines.

File: EventFilter/L1TRawToDigi/plugins/implementations_stage2/MPUnpacker.cc (hunk 3 above)
Comment (rekovic) [inline EventFilter/L1TRawToDigi/plugins/implementations_stage2/MPUnpacker.cc]: See one of comments above.  
Remove obsolete commented lines.

File: L1Trigger/Configuration/python/customiseReEmul.py (hunk 4 above)
Comment (rekovic) [inline L1Trigger/Configuration/python/customiseReEmul.py]: yes.  implemented.

Comment (rekovic) [inline L1Trigger/Configuration/python/customiseReEmul.py]: (in the next commit)

File: EventFilter/L1TRawToDigi/plugins/implementations_stage2/L1TStage2Layer2Constants.cc (hunk 2 above)
Comment (rekovic) [inline EventFilter/L1TRawToDigi/plugins/implementations_stage2/L1TStage2Layer2Constants.cc]: it should not. implemented in the next commit.

File: L1Trigger/Configuration/python/customiseReEmul.py (hunk 4 above)
Comment (rekovic) [inline L1Trigger/Configuration/python/customiseReEmul.py]: replaced with hasattr in next commit.

File: L1Trigger/L1TCalorimeter/interface/CaloParamsHelper.h (hunk 6 above)
Comment (rekovic) [inline L1Trigger/L1TCalorimeter/interface/CaloParamsHelper.h]: not necessarily, it needs modification.

File: L1Trigger/L1TCalorimeter/interface/CaloStage2Nav.h (hunk 7 above)
Comment (rekovic) [inline L1Trigger/L1TCalorimeter/interface/CaloStage2Nav.h]: simplified in the next commit.

File: L1Trigger/L1TCalorimeter/plugins/L1TStage2InputPatternWriter.cc (hunk 8 above)
Comment (rekovic) [inline L1Trigger/L1TCalorimeter/plugins/L1TStage2InputPatternWriter.cc]: This is not used in any of the workflows.  This module is used by experts when they build test files.

File: L1Trigger/L1TCalorimeter/python/caloStage2Params_2017_v1_6_2_cfi.py (hunk 9 above)
Comment (rekovic) [inline L1Trigger/L1TCalorimeter/python/caloStage2Params_2017_v1_6_2_cfi.py]: these are Calo Layer2 calibrations for L1T Jets...this is standard ....and the name is self-explanatory.

File: L1Trigger/Configuration/python/customiseSettings.py (hunk 5 above)
Comment (rekovic) [inline L1Trigger/Configuration/python/customiseSettings.py]: This is only used by the experts not in any workflow. 
But good suggestion. thanks.  will do this in the future..will certainly make the documentation shorter.

**Comment by rekovic** [comment]
please test

**Comment by rekovic** [comment]
+1

File: L1Trigger/L1TCalorimeter/python/caloStage2Params_2017_v1_6_2_cfi.py (hunk 9 above)
Comment (davidlange6) [inline L1Trigger/L1TCalorimeter/python/caloStage2Params_2017_v1_6_2_cfi.py]: yes, I know what they are.. rather a some way to know from where the numbers come?

Comment (rekovic) [inline L1Trigger/L1TCalorimeter/python/caloStage2Params_2017_v1_6_2_cfi.py]: The constants are derived using fits with MC, and usually presented at the L1T DPG meeting.

**Comment by rekovic** [comment]
+1

**Comment by rekovic** [comment]
please test

**Comment by rekovic** [comment]
@davidlange6 
please advise if anything else is needed

**Comment by davidlange6** [comment]
hi @rekovic - the large data files are already part of the externals right?

**Comment by rekovic** [comment]
hi @davidlange6  
Those 4 data files sneaked in.  Thanks for spotting it. 
Now removed them and updating the PR.
2 of them still needed to be added to externals...so made a PR to cms-data.

**Comment by rekovic** [comment]
please test

**Comment by rekovic** [comment]
please test with cms-sw/cmsdist#3118

**Comment by rekovic** [comment]
+1

**Comment by smuzaffar** [comment]
please test with cms-sw/cmsdist#3118

**Comment by rekovic** [comment]
@smuzaffar @davidlange6 

The last test I did with the command `please test with cms-sw/cmsdist#3118`
which is supposed to use the latest PR of the data externals in `L1Trigger-L1TCalorimeter`.
There were tests failing with an error that does not makes sense to me.

<pre>
----- Begin Fatal Exception 22-Jun-2017 18:48:47 CEST-----------------------
An exception of category 'Configuration' occurred while
   [0] Constructing the EventProcessor
   [1] Constructing ESProducer: class=L1TCaloStage2ParamsESProducer label='caloStage2Params'
Exception Message:
MissingParameter: Parameter 'pileUpTowerThreshold' not found.
----- End Fatal Exception -------------------------------------------------
</pre>

I cannot reproduce the problem locally with the latest IB of last night, and my tests pass. 
Here is the recipe used to test locally with IB, with an example runTheMatrix test which is successful:

<pre>
cmsrel CMSSW_9_2_X_2017-06-22-2300
cd CMSSW_9_2_X_2017-06-22-2300/src
cmsenv
git cms-merge-topic -u 19310
git clone https://github.com/cms-data/L1Trigger-L1TCalorimeter.git L1Trigger/L1TCalorimeter/data
scramv1 b -j 8
##### 
# example passing test
#####
runTheMatrix.py  -l 136.731
</pre>

@smuzaffar Unless you see something obviously wrong that I am doing, the only thing I am thinking of
is to re-run the tests.

**Comment by rekovic** [comment]
@smuzaffar 
I see you have already re-started the tests.  Our posts crossed. Thanks.

**Comment by smuzaffar** [comment]
strange, I am checking if there is any thing wrong with our PR tests script.

**Comment by smuzaffar** [comment]
@rekovic, there was bug in PR testing script for cmsdist due to which an old version ofbuild rules were being checkout and used for cmsdist PRs. I have fixed this and re-trigger the tests.

**Comment by rekovic** [comment]
@smuzaffar Thanks.

**Comment by davidlange6** [comment]
+1

**Pull request merged**